File: frontend/src/app/core/services/standard-fare.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { StandardFareRequest, StandardFareResponse } from '../../models/fare.model';

interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  timestamp: string;
}

interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}

@Injectable({
  providedIn: 'root'
})
export class StandardFareService {
  private apiUrl = `${environment.apiUrl}/standard-fares`;

  constructor(private http: HttpClient) {}

  getAllStandardFares(page: number = 0, size: number = 100): Observable<StandardFareResponse[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PageResponse<StandardFareResponse>>>(this.apiUrl, { params })
      .pipe(
        map(response => response.data.content)
      );
  }

  getStandardFareById(id: number): Observable<StandardFareResponse> {
    return this.http.get<ApiResponse<StandardFareResponse>>(`${this.apiUrl}/${id}`)
      .pipe(
        map(response => response.data)
      );
  }

  createStandardFare(request: StandardFareRequest): Observable<StandardFareResponse> {
    return this.http.post<ApiResponse<StandardFareResponse>>(this.apiUrl, request)
      .pipe(
        map(response => response.data)
      );
  }

  updateStandardFare(id: number, request: StandardFareRequest): Observable<StandardFareResponse> {
    return this.http.put<ApiResponse<StandardFareResponse>>(`${this.apiUrl}/${id}`, request)
      .pipe(
        map(response => response.data)
      );
  }

  deleteStandardFare(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`)
      .pipe(
        map(() => undefined)
      );
  }

  searchStandardFares(vehicleCode?: string, fareType?: string): Observable<StandardFareResponse[]> {
    let params = new HttpParams();
    if (vehicleCode) {
      params = params.set('vehicleCode', vehicleCode);
    }
    if (fareType) {
      params = params.set('fareType', fareType);
    }

    return this.http.get<ApiResponse<StandardFareResponse[]>>(`${this.apiUrl}/search`, { params })
      .pipe(
        map(response => response.data)
      );
  }
}

// Made with Bob

File: backend/src/main/java/com/bobcarrental/service/StandardFareService.java
package com.bobcarrental.service;

import com.bobcarrental.dto.request.StandardFareRequest;
import com.bobcarrental.dto.response.StandardFareResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for standard fare operations
 */
public interface StandardFareService {

    /**
     * Get all standard fares with pagination
     */
    Page<StandardFareResponse> getAllStandardFares(Pageable pageable);

    /**
     * Get standard fare by ID
     */
    StandardFareResponse getStandardFareById(Long id);

    /**
     * Get standard fares by vehicle type
     */
    List<StandardFareResponse> getStandardFaresByVehicleType(Long vehicleTypeId);

    /**
     * Get standard fare by vehicle type and fare type
     */
    StandardFareResponse getStandardFareByVehicleTypeAndFareType(Long vehicleTypeId, String fareType);

    /**
     * Get standard fares by fare type
     */
    List<StandardFareResponse> getStandardFaresByFareType(String fareType);

    /**
     * Get active standard fares
     */
    List<StandardFareResponse> getActiveStandardFares();

    /**
     * Create new standard fare
     */
    StandardFareResponse createStandardFare(StandardFareRequest request);

    /**
     * Update existing standard fare
     */
    StandardFareResponse updateStandardFare(Long id, StandardFareRequest request);

    /**
     * Activate standard fare
     */
    StandardFareResponse activateStandardFare(Long id);

    /**
     * Deactivate standard fare
     */
    StandardFareResponse deactivateStandardFare(Long id);

    /**
     * Delete standard fare
     */
    void deleteStandardFare(Long id);

    /**
     * Check if standard fare exists for vehicle type and fare type
     */
    boolean existsByVehicleTypeAndFareType(Long vehicleTypeId, String fareType);
}

// Made with Bob
